src/realm/object-store/sectioned_results.cpp

```diff
@@ -0,0 +1,275 @@
+    }
+}
+
+/// Returns the section of an element for the index of that element in the underlying Results.
+static SectionRange section_for_index(std::vector<SectionRange> offsets, size_t index)
+{
+    for (auto offset : offsets) {
+        if (std::find(offset.indices.begin(), offset.indices.end(), index) != offset.indices.end()) {
+            return offset;
+        }
+    }
+    throw std::logic_error("Section for given index not found.");
+}
+
+struct SectionedResultsNotificationHandler {
+public:
+    SectionedResultsNotificationHandler(SectionedResults& sectioned_results, SectionedResultsNotificatonCallback cb,
+                                        util::Optional<size_t> section_filter = util::none)
+        : m_cb(std::move(cb))
+        , m_sectioned_results(sectioned_results)
+        , m_prev_offset_ranges(m_sectioned_results.m_offset_ranges)
+        , m_section_filter(section_filter)
+    {
+    }
+
+    void before(CollectionChangeSet const&) {}
+    void after(CollectionChangeSet const& c)
+    {
+        m_sectioned_results.calculate_sections_if_required();
+        auto insertions = convert_indicies(m_sectioned_results.m_offset_ranges, c.insertions.as_indexes());
+        auto modifications = convert_indicies(m_sectioned_results.m_offset_ranges, c.modifications.as_indexes());
+        auto deletions = convert_indicies(m_prev_offset_ranges, c.deletions.as_indexes());
+
+        bool should_notify = true;
+        if (m_section_filter) {
+            bool has_insertions = insertions.count(*m_section_filter) != 0;
+            bool has_modifications = modifications.count(*m_section_filter) != 0;
+            bool has_deletions = deletions.count(*m_section_filter) != 0;
+            should_notify = has_insertions || has_modifications || has_deletions;
+        }
+        if (should_notify) {
+            m_cb(SectionedResultsChangeSet{insertions, modifications, deletions}, {});
+        }
+
+        REALM_ASSERT(m_sectioned_results.m_results.is_valid());
+        m_prev_offset_ranges = m_sectioned_results.m_offset_ranges;
+    }
+    void error(std::exception_ptr ptr)
+    {
+        m_cb({}, ptr);
+    }
+
+    /// Takes each change index and transforms it to it's corresponding section and index within that section.
+    std::map<size_t, std::vector<size_t>> convert_indicies(std::vector<SectionRange>& offsets,
+                                                           IndexSet::IndexIteratableAdaptor indicies)
+    {
+        auto modified_sections = std::map<size_t, std::vector<size_t>>();
+        for (auto i : indicies) {
+            auto range = section_for_index(offsets, i);
+            auto it = std::find(range.indices.begin(), range.indices.end(), i);
+            if (it != range.indices.end()) {
+                auto index = std::distance(range.indices.begin(), it);
+                modified_sections[range.index].push_back(index);
+            }
+        }
```

Comment:
This loop appears to be `O(N^2)`. If you have a results with N rows in a single section which are all modified, each row will do two O(N) linear scans over the array, and it won't take particularly many rows for this to start taking too long to run.

Reply:
Making this linear time could be done by calculating a bidirectional mapping when doing the sectioning. Something along the lines of:

```
    size_t size = results.size();
    std::vector<SectionRange> sections;
    std::map<Mixed, size_t> key_to_section_index;
    std::vector<std::pair<size_t, size_t>> row_to_index_path;
    row_to_index_path.resize(size);
    for (size_t i = 0; i < size; ++i) {
        Mixed key = callback(results.get_any(i), results.get_realm());
        auto it = key_to_section_index.find(key);
        if (it == key_to_section_index.end()) {
            sections.push_back({key});
            it = key_to_section_index.insert({key, sections.size() - 1}).first;
        }
        auto& section = sections[it->second];
        section.indices.push_back(i);
        row_to_index_path[i] = {it->second, section.indices.size() - 1};
    }

    auto convert_indices = [&](const IndexSet& indices) {
        std::map<size_t, IndexSet> ret;
        for (auto index : indices.as_indexes()) {
            auto& index_path = row_to_index_path[index];
            ret[index_path.first].add(index_path.second)
        }
    };

    auto insertions = convert_indicies(c.insertions);
```